### PR TITLE
`TopicCreate` Refactoring

### PR DESCRIPTION
**Pull request by joan41868**
### Description:

Addresses #461 

This **PR** modifies the `TopicCreateTransitionLogic` to use the new models.
**Defined classes**:
* `Topic` - a mutable model of a `MerkleTopic`
* `TopicStore` - responsible for interacting with the state topics (`MerkleMap`) 

**Checklist**:
- [x] Documentation - javadoc for every defined public method 
- [x] Unit tests with no less coverage than the previous implementation
- [x] Passing E2E tests for all `Topic` operations
- [x] Legacy code removal
- [x] gRPC types **only** at the top-level of the `TransitionLogic`

### Note for the reviewer:
It's good to note that I've changed one of the **negative** E2E specs for `autorenew` account. It seems that an account with `ID=1.2.3` always exists in the state, and thus - the tests fails. The account number I've introduced in the test will always result in a missing account. @Daniel-K-Ivanov

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/txns/consensus/TopicCreateTransitionLogic.java [hunk 1]
```diff
@@ -158,32 +120,20 @@ private ResponseCodeEnum validatePreSignatureValidation(TransactionBody transact
...
-			final var sanityCheck = validator.queryableAccountStatus(reqAutoRenew, accounts.get());
-			if (sanityCheck != OK) {
-				return INVALID_AUTORENEW_ACCOUNT;
-			}
-			if (ledger.isDetached(reqAutoRenew)) {
-				validationResult = ACCOUNT_EXPIRED_AND_PENDING_REMOVAL;
-			} else if (!op.hasAdminKey()) {
-				validationResult = AUTORENEW_ACCOUNT_NOT_ALLOWED;
-			}
+	private void validatePreStateTransition() {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/txns/consensus/TopicCreateTransitionLogic.java:125]: Why don't we use the precheck validation as we do in the rest of the transition logics?

File: hedera-node/src/main/java/com/hedera/services/txns/consensus/TopicCreateTransitionLogic.java [hunk 2]
```diff
@@ -158,32 +120,20 @@ private ResponseCodeEnum validatePreSignatureValidation(TransactionBody transact
...
-			final var reqAutoRenew = op.getAutoRenewAccount();
-			final var sanityCheck = validator.queryableAccountStatus(reqAutoRenew, accounts.get());
-			if (sanityCheck != OK) {
-				return INVALID_AUTORENEW_ACCOUNT;
-			}
-			if (ledger.isDetached(reqAutoRenew)) {
-				validationResult = ACCOUNT_EXPIRED_AND_PENDING_REMOVAL;
-			} else if (!op.hasAdminKey()) {
-				validationResult = AUTORENEW_ACCOUNT_NOT_ALLOWED;
-			}
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/txns/consensus/TopicCreateTransitionLogic.java:184]: I think that we should load the autorenew account and perform those checks

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]

File: hedera-node/src/main/java/com/hedera/services/txns/consensus/TopicCreateTransitionLogic.java (hunk 2 above)
Comment (joan41868) [reply hedera-node/src/main/java/com/hedera/services/txns/consensus/TopicCreateTransitionLogic.java:184]: I think this is the old implementation you commented on :)

File: hedera-node/src/main/java/com/hedera/services/txns/consensus/TopicCreateTransitionLogic.java (hunk 1 above)
Comment (joan41868) [reply hedera-node/src/main/java/com/hedera/services/txns/consensus/TopicCreateTransitionLogic.java:125]: Will do :)

Comment (joan41868) [reply hedera-node/src/main/java/com/hedera/services/txns/consensus/TopicCreateTransitionLogic.java:125]: I think there is a bit of misunderstanding here.
The `validatePreStateTransition` is actually a `post-consensus` validation, executed right before the state transition. 

The `validate` method is actually the `precheck ` validation. We will drastically change the API if we perform post-consensus validation in the precheck ( exactly 7 failing e2e specs ) .

**TL;DR:**

`TransitionLogic.validateSemantics` is the precheck, in which the API is not expecting us to validate accounts. 
`validatePreStateTransition` **is** the transition logic validation, like `validateMinting` in the `TokenMintTransitionLogic`

File: hedera-node/src/main/java/com/hedera/services/txns/consensus/TopicCreateTransitionLogic.java (hunk 2 above)
Comment (joan41868) [reply hedera-node/src/main/java/com/hedera/services/txns/consensus/TopicCreateTransitionLogic.java:184]: I think we actually have these checks performed in the current implementation: 
 `queryableAccountStatus` checks whether: 
* the account is deleted ( done in `loadAccount` )
* the account is existing (also done in `loadAccount` )
* the account is a smart contract ( explicitly done in the current implementation )

`ledger.isDetached` checks whether the account is expired, e.g. pending removal. This is also done in the `loadAcocunt()` method.
isDetached() проверява дали е акаунт и дали не е изтекъл ( става в loadAccount() )
queryableAccountStatus проверява дали съществува ( пак в loadAccount ), дали не е смарт контракт ( допълнително след loadAccount ), и дали е изтрит (пак в loadAccount )

File: hedera-node/src/main/java/com/hedera/services/txns/consensus/TopicCreateTransitionLogic.java [hunk 3]
```diff
@@ -158,32 +120,20 @@ private ResponseCodeEnum validatePreSignatureValidation(TransactionBody transact
...
+	private void validatePreStateTransition() {
+		final var op = transactionContext.accessor().getTxn().getConsensusCreateTopic();
+		final var validationResult = validator.memoCheck(op.getMemo());
+		validateTrue(OK == validationResult, validationResult);
+		validateFalse(op.hasSubmitKey() && !validator.hasGoodEncoding(op.getSubmitKey()), BAD_ENCODING);
+		validateTrue(op.hasAutoRenewPeriod(), INVALID_RENEWAL_PERIOD);
+		validateTrue(validator.isValidAutoRenewPeriod(op.getAutoRenewPeriod()), AUTORENEW_DURATION_NOT_IN_RANGE);
+		if (op.hasAutoRenewAccount()) {
+			final var autoRenew = accountStore.loadAccountOrFailWith(Id.fromGrpcAccount(op.getAutoRenewAccount()), INVALID_AUTORENEW_ACCOUNT);
+			validateFalse(autoRenew.isSmartContract(), INVALID_AUTORENEW_ACCOUNT);
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/txns/consensus/TopicCreateTransitionLogic.java:134]: We need to check whether we need to add it in the TokenUpdate as-well

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]

**Comment by joan41868** [comment]
As <a href="https://github.com/hashgraph/hedera-services/pull/2199"> this PR </a> handles the `EntityNum`'s shard/realm issue, I'll revert the changes done to the E2E spec. cc @Daniel-K-Ivanov

File: hedera-node/src/main/java/com/hedera/services/txns/consensus/TopicCreateTransitionLogic.java [hunk 4]
```diff
@@ -21,113 +21,82 @@
...
-			transactionContext.setStatus(postConsensusValidationResult);
-			return;
+		/* --- Extract gRPC --- */
+		final var transactionBody = transactionContext.accessor().getTxn();
+		final var payerAccountId = transactionBody.getTransactionID().getAccountID();
+		final var op = transactionBody.getConsensusCreateTopic();
+		/* --- Validate --- */
+		final var memoValidationResult = validator.memoCheck(op.getMemo());
+		validateTrue(OK == memoValidationResult, memoValidationResult);
+		validateFalse(op.hasSubmitKey() && !validator.hasGoodEncoding(op.getSubmitKey()), BAD_ENCODING);
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/txns/consensus/TopicCreateTransitionLogic.java:84]: We do not need to have this check here. If the submit key is not valid we will throw in the `Topic.createFromGrpc` method.

File: hedera-node/src/main/java/com/hedera/services/store/models/Topic.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+	 * Creates a new {@link Topic} from the given body.
+	 * Note: The created model is not added to state, and must be explicitly persisted via {@link com.hedera.services.store.TopicStore#persistNew(Topic)}
+	 * 
+	 * @param body - the gRPC transaction body
+	 * @param id - the id generated in the transition logic
+	 * @param expirationTime - expiration time of the topic,
+	 *               or when {@link com.hedera.services.txns.consensus.SubmitMessageTransitionLogic} will start failing.
+	 * @return - the new topic
+	 */
+	public static Topic fromGrpcTopicCreate(ConsensusCreateTopicTransactionBody body, Id id, Instant expirationTime) {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/models/Topic.java:66]: I think that it would be better to pass the variables in the method instead of passing the whole GRPC body. Here are the reasons why:
- We will cut the usage of GRPC at the TransitionLogic and not propage it in the model
- The properties are 4-5 so they will not be that many
- We can move `attemptDecodeOrThrow` to the options validator class (where I think it should be) and we can parse the keys in the transition logic from the GRPC Key.

`autoRenewAccount`, `submitKey` and `adminKey` can be `@Nullable`

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]
Overall great job! LGTM!
I have concerns about the validations around the autorenew account and more specifically, why we require `autoRenewPeriod` and why can't we have `autoRenewAccount` if there is not `adminKey`, but those we will address at the call and in the PR against Hedera as they are general inconsistencies with the autorenewal semantic and not related to the refactoring

**Review by Daniel-K-Ivanov — APPROVED** [review]
